util/src/main/java/io/kubernetes/client/informer/cache/Cache.java

```diff
@@ -0,0 +1,467 @@
+public class Cache<ApiType> implements Indexer<ApiType> {
+
+  /** keyFunc defines how to map objects into indices */
+  private KeyFunc<ApiType> keyFunc;
+
+  /** NAMESPACE_INDEX is the default index function for caching objects */
+  public static final String NAMESPACE_INDEX = "namespace";
+
+  /** indexers stores index functions by their names */
+  private Map<String, IndexFunc> indexers;
+
+  /** items stores object instances */
+  private Map<String, ApiType> items;
+
+  /** indices stores objects' keys by their indices */
+  private Map<String, Map<String, Set<String>>> indices = new HashMap<>();
+
+  /** lock provides thread-safety */
+  private ReentrantLock lock = new ReentrantLock();
+
+  /**
+   * Constructor.
+   *
+   * @param keyFunc the key func
+   * @param indexers the indexers
+   */
+  public Cache(KeyFunc keyFunc, Map<String, IndexFunc> indexers) {
+    this.keyFunc = keyFunc;
+    this.indexers = indexers;
+    this.items = new HashMap<>();
+  }
+
+  /**
+   * Add objects.
+   *
+   * @param obj the obj
+   */
+  @Override
+  public void add(ApiType obj) {
+    String key = keyFunc.getKey(obj);
+    lock.lock();
+    try {
+      ApiType oldObj = this.items.get(key);
+      this.items.put(key, obj);
+      this.updateIndices(oldObj, obj, key);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Update the object.
+   *
+   * @param obj the obj
+   */
+  @Override
+  public void update(ApiType obj) {
+    String key = keyFunc.getKey(obj);
+    lock.lock();
+    try {
+      ApiType oldObj = this.items.get(key);
+      this.items.put(key, obj);
+      updateIndices(oldObj, obj, key);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Delete the object.
+   *
+   * @param obj the obj
+   */
+  @Override
+  public void delete(ApiType obj) {
+    String key = keyFunc.getKey(obj);
+    lock.lock();
+    try {
+      boolean exists = this.items.containsKey(key);
+      if (exists) {
+        this.deleteFromIndices(this.items.get(key), key);
+        this.items.remove(key);
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Replace the content in the cache completely.
+   *
+   * @param list the list
+   * @param resourceVersion the resource version
+   */
+  @Override
+  public void replace(List<ApiType> list, String resourceVersion) {
+    lock.lock();
+    try {
+      Map<String, ApiType> newItems = new HashMap<>();
+      for (ApiType item : list) {
+        String key = keyFunc.getKey(item);
+        newItems.put(key, item);
+      }
+      this.items = newItems;
+
+      // rebuild any index
+      this.indices = new HashMap<>();
+      for (Map.Entry<String, ApiType> itemEntry : items.entrySet()) {
+        this.updateIndices(null, itemEntry.getValue(), itemEntry.getKey());
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /** Resync. */
+  @Override
+  public void resync() {
+    // Do nothing by default
```

Comment:
default

Reply:
am afraid i didn't get it. what's ur point plz?